core/lib/zksync_core/src/consensus/mod.rs

```diff
@@ -1,61 +1,212 @@
+    pub server_addr: std::net::SocketAddr,
+    /// Public address of this node (should forward to `server_addr`)
+    /// that will be advertised to peers, so that they can connect to this
+    /// node.
+    pub public_addr: std::net::SocketAddr,
+
+    /// Validator private key. Should be set only for the validator node.
+    pub validator_key: Option<SerdeText<validator::SecretKey>>,
+
+    /// Validators participating in consensus.
+    pub validator_set: Vec<SerdeText<validator::PublicKey>>,
+
+    /// Key of this node. It uniquely identifies the node.
+    pub node_key: SerdeText<node::SecretKey>,
+    /// Limit on the number of inbound connections outside
+    /// of the `static_inbound` set.
+    pub gossip_dynamic_inbound_limit: u64,
+    /// Inbound gossip connections that should be unconditionally accepted.
+    pub gossip_static_inbound: HashSet<SerdeText<node::PublicKey>>,
+    /// Outbound gossip connections that the node should actively try to
+    /// establish and maintain.
+    pub gossip_static_outbound: HashMap<SerdeText<node::PublicKey>, std::net::SocketAddr>,
+
+    pub operator_address: Option<Address>,
+}
+
+impl SerdeConfig {
+    /// Extracts consensus executor config from the `SerdeConfig`.
+    fn executor(&self) -> anyhow::Result<executor::Config> {
+        Ok(executor::Config {
+            server_addr: self.server_addr,
+            validators: validator::ValidatorSet::new(
+                self.validator_set.iter().map(|k| k.0.clone()),
+            )
+            .context("validator_set")?,
+            node_key: self.node_key.0.clone(),
+            gossip_dynamic_inbound_limit: self.gossip_dynamic_inbound_limit,
+            gossip_static_inbound: self
+                .gossip_static_inbound
+                .iter()
+                .map(|k| k.0.clone())
+                .collect(),
+            gossip_static_outbound: self
+                .gossip_static_outbound
+                .iter()
+                .map(|(k, v)| (k.0.clone(), *v))
+                .collect(),
+        })
+    }
+
+    /// Extracts a validator config from the `SerdeConfig`.
+    pub(crate) fn validator(&self) -> anyhow::Result<executor::ValidatorConfig> {
+        let key = self
+            .validator_key
+            .as_ref()
+            .context("validator_key is required")?;
+        Ok(executor::ValidatorConfig {
+            key: key.0.clone(),
+            public_addr: self.public_addr,
+        })
+    }
+}
 
-#[derive(Debug)]
+impl TryFrom<SerdeConfig> for Config {
+    type Error = anyhow::Error;
+    fn try_from(cfg: SerdeConfig) -> anyhow::Result<Self> {
+        Ok(Self {
+            executor: cfg.executor()?,
+            validator: cfg.validator()?,
+            operator_address: cfg
+                .operator_address
+                .context("operator_address is required")?,
+        })
+    }
+}
+
+/// Main node consensus config.
+#[derive(Debug, Clone)]
 pub struct Config {
```

Comment:
Bikshedding: Maybe, name this `MainNodeConfig` given that this module will presumably be used by both the main and external nodes?

Reply:
done